docs/libauthor_docs.rst

```diff
@@ -0,0 +1,70 @@
+.. currentmodule:: uarray
+
+.. _libauthordocs:
+
+Documentation for backend providers
+===================================
+
+Backend providers can provide a back-end for a defined API within
+the :obj:`uarray` ecosystem. To find out how to define your own
+API with :obj:`uarray`, see :ref:`mmauthordocs`. To find out how
+your backend will be provided, use :ref:`userdocs`.
+
+Backend providers need to be aware of three protocols: ``__ua_domain__``,
+``__ua_function__`` and ``__ua_convert__``, all of which are mandatory on
+all backends.
+
+``__ua_domain__``
+-----------------
+
+``__ua_domain__`` is a string containing the domain of the backend. This is,
+by convention, the name of the module (or one of its dependencies or parents)
+that contain the multimethods. For example, ``scipy`` and ``numpy.fft`` could
+both use the ``numpy`` domain.
+
+``__ua_function__``
+-------------------
+
+This is the most important protocol, one that defines the implementation of a
+multimethod. It has the signature ``(method, args, kwargs, dispatchable_args)``.
+Note that it is called in this form, so if your backend is an object instead of
+a module, you should add ``self``. ``method`` is the multimethod being called,
+and it is guaranteed that it is in the same domain as the backend. ``args`` and
+``kwargs`` are the arguments to the function, possibly after conversion
+(explained below). ``dispatchable_args`` are the dispatchable arguments, all
+instances of :obj:`Dispatchable`, and marked with a type.
+
+Returning :obj:`NotImplemented` signals that the backend does not support this
+operation.
+
+``__ua_convert__``
+------------------
+
+All dispatchable arguments are passed through ``__ua_convert__`` before being
+passed into ``__ua_function__``. This protocol has the signature
+``(value, dispatch_type, coerce)``. ``value`` is the value to convert,
+``dispatch_type`` is the mark of the object to be converted, and ``coerce``
+specifies whether or not to "force" the conversion. By convention, operations
+larger than ``O(log n)`` (where ``n`` is the size of the object in memory)
+should only be done if ``coerce`` is ``True``.
+
+If ``__ua_convert__`` returns ``NotImplemented``, it means that the backend in
+question does not support converting the given ``dispatch_type``, which signals
+that the value should be passed on as-is and treated as though it wasn't
+dispatchable.
```

Comment:
I still think that there needs to be a way to say "the backend doesn't support this argument". As an example: what should a CuPy backend return when passed a NumPy array and `coerce=False`?

Reply:
...of course, it could just return the NumPy array but then it would need to do redundant type-checking in its `__ua_function__`.

Reply:
How about a sentinel: `ua.TypeNotImplemented` for the current behaviour, `NotImplemented` for "I don't want to handle this, fail", and we remove `dispatchable_args` completely.

Reply:
I believe that'd satisfy all our concerns. :slightly_smiling_face:

Reply:
For your case of `ua.TypeNotImplemented`, why does this need to be a sentinel and not just have the `__ua_convert__` function pass the type through unchanged?

Reply:
Ah, yes. My original purpose was the removal from `dispatchable_args`, but that's covered already.

Reply:
This is all done, thanks for the awesome feedback!